beacon-chain/sync/pending_blocks_queue.go

```diff
@@ -132,85 +118,111 @@ func (s *Service) processPendingBlocks(ctx context.Context) error {
-				}
-			}
+// getBlocksInQueue retrieves the blocks in the pending queue for a given slot.
+func (s *Service) getBlocksInQueue(slot primitives.Slot) []interfaces.ReadOnlySignedBeaconBlock {
+	s.pendingQueueLock.RLock()
+	defer s.pendingQueueLock.RUnlock()
+	return s.pendingBlocksInCache(slot)
+}
 
-			s.pendingQueueLock.Lock()
-			if err := s.deleteBlockFromPendingQueue(slot, b, blkRoot); err != nil {
-				s.pendingQueueLock.Unlock()
-				return err
-			}
-			s.pendingQueueLock.Unlock()
+// removeBlockFromQueue removes a block from the pending queue.
+func (s *Service) removeBlockFromQueue(b interfaces.ReadOnlySignedBeaconBlock, blkRoot [32]byte) error { // Replace BlockType with the actual type
+	s.pendingQueueLock.Lock()
+	defer s.pendingQueueLock.Unlock()
+	if err := s.deleteBlockFromPendingQueue(b.Block().Slot(), b, blkRoot); err != nil {
+		return err
+	}
+	return nil
+}
 
-			log.WithFields(logrus.Fields{
-				"slot":      slot,
-				"blockRoot": hex.EncodeToString(bytesutil.Trunc(blkRoot[:])),
-			}).Debug("Processed pending block and cleared it in cache")
+// isBlockInQueue checks if a block's parent root is in the pending queue.
+func (s *Service) isBlockInQueue(parentRoot [32]byte) bool {
+	s.pendingQueueLock.RLock()
+	defer s.pendingQueueLock.RUnlock()
 
-			span.End()
+	return s.seenPendingBlocks[parentRoot]
+}
+
+// shouldRequestParentBlock determines if a parent block should be requested.
+func (s *Service) shouldRequestParentBlock(inPendingQueue, parentBlockInDB bool) bool {
+	return !inPendingQueue && !parentBlockInDB && len(s.cfg.p2p.Peers().Connected()) > 0
+}
+
+// processAndBroadcastBlock validates, processes, and broadcasts a block.
+func (s *Service) processAndBroadcastBlock(ctx context.Context, b interfaces.ReadOnlySignedBeaconBlock, blkRoot [32]byte) error { // Replace BlockType with the actual type
+	if err := s.validateBeaconBlock(ctx, b, blkRoot); err != nil {
+		if !errors.Is(ErrOptimisticParent, err) {
+			log.WithError(err).WithField("slot", b.Block().Slot()).Debug("Could not validate block")
+			return err
 		}
 	}
 
-	return s.sendBatchRootRequest(ctx, parentRoots, randGen)
+	if err := s.cfg.chain.ReceiveBlock(ctx, b, blkRoot); err != nil {
+		return err
+	}
+
+	s.setSeenBlockIndexSlot(b.Block().Slot(), b.Block().ProposerIndex())
+
+	pb, err := b.Proto()
+	if err != nil {
+		log.WithError(err).Debug("Could not get protobuf block")
+		return err
+	}
+	if err := s.cfg.p2p.Broadcast(ctx, pb); err != nil {
+		log.WithError(err).Debug("Could not broadcast block")
+		return err
+	}
+
+	return nil
+}
+
+// handleBlockProcessingError handles errors during block processing.
+func (s *Service) handleBlockProcessingError(ctx context.Context, err error, b interfaces.ReadOnlySignedBeaconBlock, blkRoot [32]byte) { // Replace BlockType with the actual type
+	if blockchain.IsInvalidBlock(err) {
+		r := blockchain.InvalidBlockRoot(err)
+		if r != [32]byte{} {
+			s.setBadBlock(ctx, r)
+		} else {
+			s.setBadBlock(ctx, blkRoot)
```

Comment:
This is gross, what's the point of having a typed error with a special accessor for an embedded value if that value can sometimes get set to something goofed up?

Reply:
fixed in aa501fc21e5a12b63514053232f0d217d6e82497